src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs

```diff
@@ -0,0 +1,153 @@
+
+			context.RegisterSyntaxNodeAction(
+				ctx => {
+					if( ctx.Node is ObjectCreationExpressionSyntax expr ) {
+						AnalyzeInvocation( ctx, expr );
+					}
+				},
+				SyntaxKind.ObjectCreationExpression
+			);
+		}
+
+		private void AnalyzeInvocation(
+			SyntaxNodeAnalysisContext context,
+			ObjectCreationExpressionSyntax syntax
+		) {
+
+			TypeSyntax type = syntax.Type;
+			ISymbol symbol = context
+				.SemanticModel
+				.GetSymbolInfo( type ).Symbol;
+
+			if( symbol == null ) {
+				return;
+			}
+
+			if( !IsStatelessFunc( symbol ) ) {
+				return;
+			}
+
+			ExpressionSyntax argument = syntax.ArgumentList.Arguments[ 0 ].Expression;
+			SyntaxKind kind = argument.Kind();
+			switch( kind ) {
+
+				// this is the case when a method reference is used
+				// eg Func<string, int> func = int.Parse
+				case SyntaxKind.SimpleMemberAccessExpression:
+					if( IsStaticMemberAccess( context, argument ) ) {
+						return;
+					}
+
+					// non-static member access means that state could
+					// be used / held.
+					// TODO: Look for [Immutable] on the member's type
+					// to determine if the non-static member is safe
+					break;
+
+				// this is the case when the left hand side of the
+				// lambda has parens
+				// eg () => 1, (x, y) => x + y
+				case SyntaxKind.ParenthesizedLambdaExpression:
+					if( !CheckForClosures( context, argument ) ) {
+						return;
+					}
+					break;
+
+				// this is the case when the left hand side of the
+				// lambda does not have parens
+				// eg x => x + 1
+				case SyntaxKind.SimpleLambdaExpression:
+					if( !CheckForClosures( context, argument ) ) {
+						return;
+					}
+					break;
+
+				default:
+					// should we do something else here because constructors
+					// of D2L.StatelessFunc<T> and friends should always take
+					// a func?
+					return;
+			}
+
+			var diag = Diagnostic.Create(
+				Diagnostics.StatelessFuncIsnt,
+				argument.GetLocation()
+			);
+
+			context.ReportDiagnostic( diag );
+		}
+
+		private static bool IsStaticMemberAccess(
+			SyntaxNodeAnalysisContext context,
+			ExpressionSyntax expression
+		) {
+			ISymbol symbol = context
+				.SemanticModel
+				.GetSymbolInfo( expression ).Symbol;
+
+			if( symbol == null ) {
+				return false;
+			}
+
+			return symbol.IsStatic;
+		}
+
+		private static bool CheckForClosures(
+			SyntaxNodeAnalysisContext context,
+			ExpressionSyntax expression
+		) {
+
+			DataFlowAnalysis dataFlow = context
+				.SemanticModel
+				.AnalyzeDataFlow( expression );
+
+			ImmutableArray<ISymbol> captures = dataFlow.Captured;
+			return captures.Length > 0;
+		}
+
+		private static bool IsStatelessFunc( ISymbol symbol ) {
+			ImmutableHashSet<string> statelessFuncs = GetStatelessFuncTypes();
+
+			string fullName = $"{ symbol.ContainingNamespace.Name }.{ symbol.MetadataName }";
+
+			return statelessFuncs.Contains( fullName );
+		}
+
+		private static ImmutableHashSet<string> GetStatelessFuncTypes() {
```

Comment:
Could this just become a `private static readonly ImmutableHashSet<string>` field/property? Rather than the builder etc.